File: insert_digit_anywhere_primes.py
import math

class Tree:
    def __init__(self, value, children):
        self.value = value
        self.children = []
        for child in children:
            self.children.append(Tree(child, []))

    def step(self):
        if len(self.children) == 0:
            for val in step(self.value):
                self.children.append(Tree(val, []))
            return
        for child in self.children:
            child.step()

    def __str__(self):
        retval = '[' + str(self.value) + ': '
        for child in self.children:
            retval += str(child)
        retval += ']'
        return retval

    def __repr__(self):
        return str(self)

    def longest_path(self):
        if len(self.children) == 0:
            return [self.value]
        max_path = []
        max_length = 0
        for child in self.children:
            temp = child.longest_path()
            if len(temp) > max_length:
                max_path = temp
                max_length = len(temp)
        return [self.value] + max_path

def is_square(x):
    return int(math.sqrt(x)) == math.sqrt(x)

primes = {}

def is_prime(x):
    try:
        return primes[x]
    except KeyError:
        i = 2
        while i * i <= x:
            if x % i == 0:
                primes[x] = False
                return False
            i += 1
        primes[x] = True
        return True

def step(x):
    new_xs = []
    str_x = str(x)
    for (i,digit) in enumerate(str_x + '0'):
        for d in range(0,10):
            temp = str_x[:i] + str(d) + str_x[i:]
            if not (i == 0 and d == 0):
                if is_prime(int(temp)):
                    new_xs.append(int(temp))
    return new_xs


def main():
    tree = Tree(0, [2, 3, 5, 7])
    for i in range(20):
        print(tree.longest_path())
        tree.step()
    print(tree.longest_path())


if __name__ == '__main__':
    main()

File: insert_digit_anywhere_primes.rs
use std::collections::HashMap;

#[derive(Debug)]
struct Tree {
    value: u64,
    children: Vec<Tree>,
}

impl Tree {
    fn new(value: u64, children: Vec<u64>) -> Tree {
        let mut tree_children = Vec::with_capacity(children.len());
        for child in children {
            tree_children.push(Tree::new(child, Vec::new()));
        }
        Tree {
            value,
            children: tree_children,
        }
    }

    fn step(&mut self, primes: &mut HashMap<u64, bool>) {
        if self.children.is_empty() {
            let str_x = self.value.to_string();
            for i in 0..str_x.len() + 1 {
                for d in 0..10 {
                    let temp = format!("{}{}{}", &str_x[..i], d.to_string(), &str_x[i..]).parse().unwrap();
                    if !(i == 0 && d == 0) {
                        if is_prime(temp, primes) {
                            self.children.push(Tree::new(temp, Vec::new()));
                        }
                    }
                }
            }
            return;
        }
        for child in &mut self.children {
            child.step(primes);
        }
    }

    fn longest_path(&self) -> Vec<u64> {
        if self.children.is_empty() {
            return vec![self.value];
        }
        let mut max_path = vec![];
        let mut max_length = 0;
        for child in &self.children {
            let temp = child.longest_path();
            if temp.len() > max_length {
                max_length = temp.len();
                max_path = temp;
            }
        }
        let mut retval = vec![self.value];
        retval.append(&mut max_path);
        return retval;
    }
}

fn is_prime(x: u64, primes: &mut HashMap<u64, bool>) -> bool {
    let is_prime = primes.get(&x);
    match is_prime {
        Some(val) => return *val,
        None => {
            let mut i = 2;
            while i * i <= x {
                if x % i == 0 {
                    primes.insert(x, false);
                    return false;
                }
                i += 1;
            }
            primes.insert(x, true);
            return true;
        },
    }
}

fn main() {
    let mut tree = Tree::new(0, vec![2, 3, 5, 7]);
    let mut primes: HashMap<u64, bool> = HashMap::new();
    for _ in 0..10 {
        tree.step(&mut primes);
        println!("{:?}", tree.longest_path());
    }
}
